### ISSUE TITLE
Proposal: c10d ProcessGroup extension and C++ API change

### ISSUE THREAD

**Issue by agolynski** [pytorch/pytorch#39662]
# Proposal: c10d ProcessGroup extension and C++ API change

## Purpose

Clients to provide custom implementations of ProcessGroup API and use them as CCL backend without needing to modify PyTorch codebase. Registration mechanism is implemented in https://github.com/pytorch/pytorch/pull/28068.

### Usage example

Implement constructor of a single rank in your custom ProcessGroup called `custom_backend` implemented in c++ module `custom_backend_cpp_module (c++):`

Register your backend (c++, pybind):

```
`dist.Backend.register_backend("**custom_backend**``", py::cpp_function(`**`GetNewRank`**`));`
```

Load your module (py):

```
`torch``.``utils``.``cpp_extension``.``load``(``name``=``"**custom_backend_cpp_module**``"``,`` sources``=[...])`
```

Initialize process group (py):

```
`dist``.``init_process_group``(``backend``=``"**custom_backend**``"``,`` init_method ``=`` ``...)`
```

## Proposed changes to the current API

ProcessGroup API has some inconsistencies, so we’d like to make it more uniform across different functionalities.

For some operations like allgather and reducescatter we allow in-place mode by means of user can specifying in_place = true in options of the corresponding operation. In this mode, user guarantees that inputs are unchanged until the operation completes. Implementors of ProcessGroup can (and are encouraged to) take advantage of this mode to speed up their implementations, e.g. see https://github.com/pytorch/pytorch/pull/33924.

Some operations have multi-gpu versions:
reduce, allreduce, gather where a single process group controls multiple GPUs. These are mostly used by data parallel. [Should we mention: For newer applications, we suggest to have one rank of process group per GPU?]

What we are changing in new API to make it more straightforward:

1) Batched operations (main purpose is to add efficiency by combining tensors) are going to live outside of ProcessGroup API so there is no need to implement them by 3rd party:
allreduce_coalesced, allgather_coalesced

2) adding single-GPU versions of allreduce, gather, scatter

3) removing multi-GPU versions of scatter, gather that aren’t supported by any of existing backends.

## All to all

Current API (**keeping**). 

```
  virtual std::shared_ptr<ProcessGroup::Work> alltoall(
      std::vector<at::Tensor>& outputTensors,
      std::vector<at::Tensor>& inputTensors,
      const AllToAllOptions& opts = AllToAllOptions()) = 0;
```

Current support: MPI only. (Clients are encouraged to do optimizations based on tensor layout, i.e. not making extra tensor copies if tensors share storage and properly aligned)

Current API (**removing**):

```
  virtual std::shared_ptr<ProcessGroup::Work> alltoall_base(
      at::Tensor& outputTensor,
      at::Tensor& inputTensor,
      std::vector<int64_t>& outputSplitSizes,
      std::vector<int64_t>& inputSplitSizes,
      const AllToAllOptions& opts = AllToAllOptions()) = 0;
```

## All reduce

Current API (keeping until status of multigpu PG is clear):

```
allreduce(std::vector<at::Tensor>& data, const AllreduceOptions& opts) = 0;
```

*Current support*: coalesced version is supported in GLOO only.

**Adding** single tensor API:

```
// Not inlcuded in API at the moment: allreduce a single tensor.
allreduce(at::Tensor& tensor, const AllreduceOptions& opts) = 0;
```

**No longer a part of API: all reduce coalesced, it will be moved away from process group (no need for 3rd parties to implement) and included in comms instead.**

```
// Being deprecated and moved to comms, can be done via allreduce_base.
allreduce_coalesced(std::vector<at::Tensor>& tensors, 
                    const AllreduceCoalescedOptions& opts);
```

## Gather

Current API (**not keeping**):

```
gather(std::vector<std::vector<at::Tensor>>& outputTensors,
       std::vector<at::Tensor>& inputTensors,
       const GatherOptions& opts) = 0;
```

*Current support: *MPI, GLOO only supports single element gather, NCCL does not support gather. 

Desirable API:

```
gather(std::vector<at::Tensor>& outputTensors,
       at::Tensor& inputTensor,
       const GatherOptions& opts) = 0;
```

## All gather

Current API:

```
  virtual std::shared_ptr<ProcessGroup::Work> allgather(
      std::vector<std::vector<at::Tensor>>& outputTensors,
      std::vector<at::Tensor>& inputTensors,
      const AllgatherOptions& opts = AllgatherOptions())= 0;
```

*Current support: *coalesced version is supported in GLOO only*.*

Desirable API (**adding**) 

```
  virtual std::shared_ptr<ProcessGroup::Work> allgather(
      std::vector<at::Tensor>& outputTensors,
      at::Tensor& inputTensor,
      const AllgatherOptions& opts = AllgatherOptions()) = 0;
```

(Clients are encouraged to do optimizations based on tensor layout, i.e. not making extra tensor copies if tensors share storage and properly aligned)

**Not part of API (same as with allreduce_coalesced)**

```
  // Being deprecated and moved to comms, can be done via allgather_base.
  virtual std::shared_ptr<ProcessGroup::Work> allgather_coalesced(
      std::vector<std::vector<at::Tensor>>& outputTensorLists,
      std::vector<at::Tensor>& inputTensors,
      const AllgatherOptions& opts = AllgatherOptions())ß;
```

## Scatter

Current API (**not keeping**):

```
  virtual std::shared_ptr<ProcessGroup::Work> scatter(
      std::vector<at::Tensor>& outputTensors,
      std::vector<std::vector<at::Tensor>>& inputTensors,
      const ScatterOptions& opts = ScatterOptions()) = 0;
```

*Current support: *GLOO, MPI supports single tensor version only, NCCL does not support scatter. 

Desired API (**adding**):

```
  virtual std::shared_ptr<ProcessGroup::Work> scatter(
      at::Tensor& outputTensor,
      std::vector<at::Tensor>& inputTensors,
      const ScatterOptions& opts = ScatterOptions()) = 0;
```

### Reduce

```
reduce(std::vector<at::Tensor>& tensors, const ReduceOptions& opts) = 0;
```

*Current support*: multiple tensors for usecase where ‘tensors’ should live on different GPUs, GLOO, MPI supports single tensor input only. NCCL supports multiple tensors.

### ReduceScatter

Current API:

```
  virtual std::shared_ptr<ProcessGroup::Work> reduce_scatter(
      std::vector<at::Tensor>& outputTensors,
      std::vector<std::vector<at::Tensor>>& inputTensors,
      const ReduceScatterOptions& opts = ReduceScatterOptions()) = 0;
```

*Current support: *NCCL only*.*

### Broadcast

```
  virtual std::shared_ptr<ProcessGroup::Work> broadcast(
      std::vector<at::Tensor>& data,
      const BroadcastOptions& opts = BroadcastOptions()) = 0;
```

### Barrier

```
barrier(const BarrierOptions& opts) = 0;
```

### Send/receive

```
  virtual std::shared_ptr<ProcessGroup::Work> send(
      std::vector<at::Tensor>& tensors,
      int dstRank,
      int tag) = 0;

  virtual std::shared_ptr<ProcessGroup::Work> recv(
      std::vector<at::Tensor>& tensors,
      int srcRank,
      int tag) = 0;

  virtual std::shared_ptr<ProcessGroup::Work> recvAnysource(
      std::vector<at::Tensor>& tensors,
      int tag) = 0;
```

## Backend/feature support table

| Backend        | ``gloo``|   | ``mpi``|   | ``nccl``  | |
|----------------|-----|-----|-----|-----|-----|-----|
| Device         | CPU | GPU | CPU | GPU | CPU | GPU |
| send           | ✓   | ✘   | ✓   | ?   | ✘   | ✘   |
| recv           | ✓   | ✘   | ✓   | ?   | ✘   | ✘   |
| broadcast      | ✓   | ✓   | ✓   | ?   | ✘   | ✓   |
| all_reduce     | ✓   | ✓   | ✓   | ?   | ✘   | ✓   |
| reduce         | ✓   | ✘   | ✓   | ?   | ✘   | ✓   |
| all_gather     | ✓   | ✘   | ✓   | ?   | ✘   | ✓   |
| gather         | ✓   | ✘   | ✓   | ?   | ✘   | ✘   |
| scatter        | ✓   | ✘   | ✓   | ?   | ✘   | ✘   |
| reduce_scatter | ✘   | ✘   | ✘   | ✘   | ✘   | ✓   |
| all_to_all     | ✘   | ✘   | ✓   | ?   | ✘   | ✘   |
| barrier        | ✓   | ✘   | ✓   | ?   | ✘   | ✓   |

**Comment by mshiryaev** [comment]
@agolynski, could you please clarify few questions?

- > Some operations have multi-gpu versions: reduce, allreduce, gather where a single process group controls multiple GPUs. These are mostly used by data parallel. [Should we mention: For newer applications, we suggest to have one rank of process group per GPU?]

Just wonder what is the main motivation to remove multi-gpu methods from ProcessGroup (PG) API?

As far as I understand currently user can create 1 process with 1 PG object with desired rank and use this object for multi-gpu communications.

When multi-gpu method is called in NCCL PG then it expands user-supplied rank into multiple NCCL ranks (https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L505).
In case of Gloo PG doesn't expand user-supplied rank and pass vector of tensors as is into Gloo where they are handled properly (https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupGloo.cpp#L531).

What will be new suggested approach to connect multiple GPUs?

1. Should user create 1 process with 1 PG object which maps on single GPU, i.e. single process managing single PG rank?
2. Or should user create 1 process with multiple PG objects each owning single GPU, i.e. single process managing multiple PG ranks?

---

- > Batched operations (main purpose is to add efficiency by combining tensors) are going to live outside of ProcessGroup API so there is no need to implement them by 3rd party: allreduce_coalesced, allgather_coalesced

Not directly related with "*_coalesced" methods but still relevant question - can you consider to add knob to disable batching which happens on c10d::Reducer level?
Transport level may implement own batching mechanism and it may be beneficial to propagate independent collective operations as is from upper level to PG level.

---

- > Current API (removing):
  virtual std::shared_ptr<ProcessGroup::Work> alltoall_base(
  
Do you consider to remove vector alltoall support? What is the motivation? @ddkalamk, FYI.

---

- Each collective operation in PG accepts \<coll\>Options parameter. What is about to create separate BaseOptions structure with options which are applicable for all collectives and inherit collective specific structures from this base one?
Currently "timeout" can be moved into BasicOptions. And what is about to extend BasicOptions with such option like "async" (to skip creation of PG::Work object when async=0) and further with other common options.

- Can you consider to add ability to retrieve tensor name from at::Tensor object or pass it as independent parameter, or along with \<coll\>Options into PG operations? It can be used by transport level for context-specific features.

**Comment by mshiryaev** [comment]
>> The main motivation is clean up for methods that we never implemented, e.g. scatter that is either not implemented in backends or only supported for one GPU...
We are leaving multi-GPU versions of other collective primitives that are currently supported. We are thinking to recommend users to use one GPU per PG, but that's outside of scope of this proposal.

Thanks for clarification.

---

>> Let me clarify on transport level batching mechanism: what are examples of such lower level batching on existing backends?

I don't have example on existing built-in backends (Gloo, NCCL, MPI). But other backends like CCL may make own decision when batching should happen (according to own thresholds, timers, parallelization strategy). c10d::Reducer makes batching in single master thread only and according to own buckets only. Transport level can try to optimize this. Also for CPU collectives we noticed that c10d::Reducer batching may add overhead vs execution of each collective independently.
All this also outside of scope of original proposal, but just like idea for consideration.

---

>> The motivation is to make API more straightforward, the APIs should be functionally equivalent ...

I see, thanks. So alltoallv is still possible by providing differently sized input/output tensors in PG::alltoall. And we still can check whether tensors share the same storage and make corresponding optimization. This may affect our use-case, but we still can handle it.

---

>> That's a good idea I think. Do we have any other data pieces to extract in common class?

You may also consider to add into BaseOptions such fields like "priority" - for user-defined priority of collective, "prologue", "epilogue" - for user-defined operations before and after collective operation, to be used for example for compress/decompress logic fused with collective itself.

---

>> Apologies, I am not quite following your proposal, could you give an example?

The proposal is to adjust at::Tensor to contain some unique identifier (like name string) which can be passed to transport level to distinguish different collective operations and makes context-aware logic, like caching of internal structures for specific tensor collective.

PyTorch model may build unique identifiers but they can't accessed from tensor class.
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L919

Passing of names through \<coll\>Options structure doesn't look feasible because PG::collective may accept multiple input/output tensors but only single \<\coll>Options parameter.

**Comment by skyw** [comment]
The tensor list input of alltoall is useful in some cases. For example, in recommender training, the all to all in backward propagation is transposed w.r.t forward propagation, which means the "offset" on dimension in dimension 1 instead of 0.

The code snippet I use like this
```python
grad_local_bottom_outputs_split = list(grad_local_bottom_outputs.split(ctx.batch_size_per_gpu, dim=0))

split_grads = [t.contiguous() for t  in (grad_slice_bottom_outputs.view(ctx.batch_size_per_gpu, -1).split(
    [ctx.vector_dim * n for n in ctx.vectors_per_gpu], dim=1))]

torch.distributed.all_to_all(grad_local_bottom_outputs_split, split_grads)
```
To use the single tensor input all to all API, it will need to copy tensor list split by dim=1 to another big tensor. And another contiguous call on the big tensor maybe required in additional to the copy.

**Comment by agolynski** [comment]
> The tensor list input of alltoall is useful in some cases. For example, in recommender training, the all to all in backward propagation is transposed w.r.t forward propagation, which means the "offset" on dimension in dimension 1 instead of 0.
> 
> The code snippet I use like this
> 
> ```python
> grad_local_bottom_outputs_split = list(grad_local_bottom_outputs.split(ctx.batch_size_per_gpu, dim=0))
> 
> split_grads = [t.contiguous() for t  in (grad_slice_bottom_outputs.view(ctx.batch_size_per_gpu, -1).split(
>     [ctx.vector_dim * n for n in ctx.vectors_per_gpu], dim=1))]
> 
> torch.distributed.all_to_all(grad_local_bottom_outputs_split, split_grads)
> ```
> 
> To use the single tensor input all to all API, it will need to copy tensor list split by dim=1 to another big tensor. And another contiguous call on the big tensor maybe required in additional to the copy.

Just to make sure, we propose the list input in this proposal instead of one input/output tensor with sizes.

One thing to keep in mind: the nature of implementation for this operation is such that if tensors aren't sharing same storage or aren't properly aligned within it, we'd have to put them into one contiguous tensor behind the scenes.

**Comment by msbaines** [comment]
I think there are issues with the c10d extension API. List[ranks] is not passed when creating a new group. MPI requires this. I worked around by replacing the builtin MPI. Also, MPI requires all ranks participate when creating a new group. The current implementation only calls the create method on the ranks that are part of the new group. In summary, it is not possible to "cleanly" implement ProcessGroupMPI using the current c10d extension mechanism.

Here is my external impl of c10d extension of MPI and NCCL:

https://github.com/msbaines/torch_pg

My hacks are here:

https://github.com/msbaines/torch_pg/blob/main/torch_pg/__init__.py

My original plan was to call these "mpi-ext" and "nccl-ext" but due to the issues above I had to take over "mpi".
